main/src/com/google/refine/operations/row/RowDuplicatesRemovalOperation.java

```diff
@@ -0,0 +1,130 @@
+import com.google.refine.browsing.RowVisitor;
+import com.google.refine.history.HistoryEntry;
+import com.google.refine.model.Column;
+import com.google.refine.model.Project;
+import com.google.refine.model.Row;
+import com.google.refine.model.changes.RowRemovalChange;
+import com.google.refine.operations.EngineDependentOperation;
+import com.google.refine.operations.OperationDescription;
+
+public class RowDuplicatesRemovalOperation extends EngineDependentOperation {
+
+    final protected List<String> _criteria;
+    final List<Column> criteriaColumns = new ArrayList<Column>();
+
+    @JsonCreator
+    public RowDuplicatesRemovalOperation(
+            @JsonProperty("engineConfig") EngineConfig engineConfig,
+            @JsonProperty("criteria") List<String> criteria) {
+        super(engineConfig);
+        _criteria = criteria;
+    }
+
+    @JsonProperty("criteria")
+    public List<String> getRows() {
+        return _criteria;
+    }
+
+    @Override
+    protected String getBriefDescription(Project project) {
+        return OperationDescription.row_remove_duplicates_brief();
+    }
+
+    @Override
+    protected HistoryEntry createHistoryEntry(Project project, long historyEntryID) throws Exception {
+        Engine engine = createEngine(project);
+
+        List<Integer> rowIndices = new ArrayList<Integer>();
+        for (String c : _criteria) {
+            Column toColumn = project.columnModel.getColumnByName(c);
+            if (toColumn != null) {
+                criteriaColumns.add(toColumn);
+            }
+        }
+
+        FilteredRows filteredRows = engine.getAllFilteredRows();
+        filteredRows.accept(project, createRowVisitor(project, rowIndices));
+
+        return new HistoryEntry(
+                historyEntryID,
+                project,
+                "Remove " + rowIndices.size() + " rows",
+                this,
+                new RowRemovalChange(rowIndices));
+    }
+
+    protected RowVisitor createRowVisitor(Project project, List<Integer> rowIndices) throws Exception {
+        return new RowVisitor() {
+
+            List<Integer> rowIndices;
+            HashSet<Object> rowUniqueKeys = new HashSet<>();
+
+            public RowVisitor init(List<Integer> rowIndices) {
+                this.rowIndices = rowIndices;
+                return this;
+            }
+
+            @Override
+            public void start(Project project) {
+                // nothing to do
+            }
+
+            @Override
+            public void end(Project project) {
+                // nothing to do
+            }
+
+            @Override
+            public boolean visit(Project project, int rowIndex, Row row) {
+                if (isDuplicate(row, rowUniqueKeys)) {
+                    rowIndices.add(rowIndex);
+                }
+                return false;
+            }
+        }.init(rowIndices);
+    }
+
+    private boolean isDuplicate(Row row, HashSet<Object> rowUniqueKeys) {
```

Comment:
You could even move that method inside the visitor, which would mean you wouldn't even need to pass the `rowUniqueKeys` as parameter as it would already be accessible as an attribute.

Reply:
Addressed